lua/lazyvim/plugins/extras/lang/java.lua

```diff
@@ -24,102 +24,134 @@ return {
-                table.insert(bundles, bundle)
-              end
-            end
-          end
-          -- Attach jdtls for the proper filetypes (i.e. java).
-          -- Existing server will be reused if the root_dir matches.
-          vim.api.nvim_create_autocmd("FileType", {
-            group = vim.api.nvim_create_augroup("MyJdtls", { clear = true }),
-            pattern = filetypes,
-            callback = function()
-              local fname = vim.api.nvim_buf_get_name(0)
-              local root_dir = find_java_project_root(fname)
-              local project_name = root_dir and vim.fs.basename(root_dir)
-              local cmd = { "jdtls" }
-              if project_name then
-                local jdtls_cache_dir = vim.fn.stdpath("cache") .. "/jdtls/" .. project_name
-                vim.list_extend(cmd, {
-                  "-configuration",
-                  jdtls_cache_dir .. "/config",
-                  "-data",
-                  jdtls_cache_dir .. "/workspace",
-                })
-              end
-              local jdtls_base_config = {
-                on_attach = require("lazyvim.util").on_attach(function(client, buffer)
-                  if mason_registry.has_package("java-test") then
-                    -- custom keymaps for Java test runner (not yet compatible with neotest)
-                    vim.keymap.set("n", "<leader>tT", function() require("jdtls").pick_test({ bufnr = buffer }) end, { buffer = buffer, desc = "Run specific Test" })
-                    vim.keymap.set("n", "<leader>tt", function() require("jdtls").test_class({ bufnr = buffer }) end, { buffer = buffer, desc = "Run File" })
-                    vim.keymap.set("n", "<leader>tr", function() require("jdtls").test_nearest_method({ bufnr = buffer }) end, { buffer = buffer, desc = "Run nearest" })
-                  end
-                  if mason_registry.has_package("java-debug-adapter") then
-                    -- custom init for Java debugger
-                    require("jdtls").setup_dap({ hotcodereplace = "auto" })
-                    require("jdtls.dap").setup_dap_main_class_configs()
-                  end
-                  require("jdtls.setup").add_commands()
-                end),
-                cmd = cmd,
-                root_dir = root_dir,
-                init_options = {
-                  bundles = bundles,
-                }
-              }
-              local jdtls_opts = require("lazyvim.util").opts("nvim-jdtls")
-              require("jdtls").start_or_attach(vim.tbl_deep_extend("force", jdtls_opts or {}, jdtls_base_config))
-              require("which-key").register({ c = { x = { name = "Extract" } } }, { prefix = "<leader>" })
-            end,
-          })
           return true -- avoid duplicate servers
         end,
       },
     },
   },
+
+  -- Set up nvim-jdtls
+  {
+    "mfussenegger/nvim-jdtls",
+    dependencies = { "folke/which-key.nvim" },
+    ft = { "java" },
```

Comment:
This `ft = { "java" }` needs to be omitted, because the config function also sets up filetype-based autocmds, and the new ones don't fire properly if loading depends on the filetype.

Reply:
Plz take what I'm about to say with a grain of salt, since I don't know anything about Java, I'm just commenting based on my experience with `lazy.nvim`. Couldn't you maybe keep the `ft = { "java" }` and then load the autocmds in the config function based on some event other than `FileType`? For example something like `BufReadPre(Post)`, since `jdtls` will load on `java` filetypes either way? Have you tested such a possibility or is this not going to work? If it works this way, then you will also not have the problem of `jdtls` loading when not in Java files.

Reply:
@dpetka2001 Thanks for your comment, and my apologies for not replying earlier.

We'll keep `ft = { "java" }` to preserve lazy loading, but then handle the first-load case specially. See #1388